RecoTracker/TkTrackingRegions/plugins/CandidatePointSeededTrackingRegionsProducer.h

```diff
@@ -0,0 +1,456 @@
+
+      for(const auto& object : objs) {
+
+	double eta_Cand = object.eta();
+	double phi_Cand = object.phi();
+		
+	for( const auto& directionPoint : directionPoints ){
+
+	  double eta_Point = directionPoint.eta();
+	  double phi_Point = directionPoint.phi();
+	  double dEta_Cand_Point = std::abs(eta_Cand-eta_Point);
+	  double dPhi_Cand_Point = std::abs(deltaPhi(phi_Cand,phi_Point));
+
+	  //Check if there is an overlap between Candidate- and Point-based regions of interest
+	  if(dEta_Cand_Point > (m_deltaEta_Cand + m_deltaEta_Point) || dPhi_Cand_Point > (m_deltaPhi_Cand + m_deltaPhi_Point)) continue;
+
+	  //Determines boundaries of intersection of RoIs
+	  double etaMin_RoI = std::max(eta_Cand-m_deltaEta_Cand,eta_Point-m_deltaEta_Point);
+	  double etaMax_RoI = std::min(eta_Cand+m_deltaEta_Cand,eta_Point+m_deltaEta_Point);
+
+	  double phi_Cand_minus  = normalizedPhi(phi_Cand-m_deltaPhi_Cand);
+	  double phi_Point_minus = normalizedPhi(phi_Point-m_deltaPhi_Point);
+	  double phi_Cand_plus  = normalizedPhi(phi_Cand+m_deltaPhi_Cand);
+	  double phi_Point_plus = normalizedPhi(phi_Point+m_deltaPhi_Point);
+
+	  double phiMin_RoI = deltaPhi(phi_Cand_minus,phi_Point_minus)>0. ? phi_Cand_minus : phi_Point_minus ;	
+	  double phiMax_RoI = deltaPhi(phi_Cand_plus,phi_Point_plus)<0. ? phi_Cand_plus : phi_Point_plus;
+
+	  //Determines position and width of new RoI
+	  double eta_RoI = 0.5*(etaMax_RoI+etaMin_RoI);
+	  double deltaEta_RoI = etaMax_RoI - eta_RoI;
+
+	  double phi_RoI = 0.5*(phiMax_RoI+phiMin_RoI);
+	  if( phiMax_RoI < phiMin_RoI ) phi_RoI-=M_PI;
+	  phi_RoI = normalizedPhi(phi_RoI);
+	  double deltaPhi_RoI = deltaPhi(phiMax_RoI,phi_RoI);
+	  
+	  double x = std::cos(phi_RoI);
+	  double y = std::sin(phi_RoI);
+	  double theta = 2*std::atan(std::exp(-eta_RoI));
+	  double z = 1./std::tan(theta);
+
+	  GlobalVector direction( x,y,z );		
+      
+	  for(const auto& origin : origins) {
+	      
+	    result.push_back(std::make_unique<RectangularEtaPhiTrackingRegion>(
+		direction,
+		origin.first,
+		m_ptMin,
+		m_originRadius,
+		origin.second,
+		deltaEta_RoI,
+		deltaPhi_RoI,
+		m_whereToUseMeasurementTracker,
+		m_precise,
+		measurementTracker,
+		m_searchOpt
+	        ));	        
+	    ++n_regions;
+	    if( n_regions >= m_maxNRegions ) break;
+
+	  }
+
+	  if( n_regions >= m_maxNRegions ) break;
+
+	}
+
+	if( n_regions >= m_maxNRegions ) break;
+	
+      }
+
+    }
+    
+    edm::LogInfo ("CandidatePointSeededTrackingRegionsProducer") << "produced "<<n_regions<<" regions";
+
+    return result;
+  }
+  
+private:
+
+  OperationMode m_operationMode;
+  SeedingMode m_seedingMode;
+
+  int m_maxNRegions;
+  edm::EDGetTokenT<reco::VertexCollection> token_vertex; 
+  edm::EDGetTokenT<reco::BeamSpot> token_beamSpot; 
+  edm::EDGetTokenT<reco::CandidateView> token_input; 
+  int m_maxNVertices;
+
+  size_t n_points;
+  std::vector<double> etaPoints;
+  std::vector<double> phiPoints;
+  std::vector<GlobalVector> directionPoints;
```

Comment:
hi @tstreble - it seems to me that its more natural to store pairs of phi and eta rather than to repeat the conversion to/from a GlobalVector (which is needed in one direction thanks to the interface of RectangularEtaPhiTrackingRegion)

Reply:
Sorry I am not sure I fully understand your comment
As you pointed out, a GlobalVector is needed as input of the RectangularEtaPhiTrackingRegion.
It has been pointed out earlier in the review that it would probably be more optimal to have the computations needed for this conversion in the constructor, instead of repeating them for every event.
Am I missing something or do you have something specific in mind?